pdns/syncres.cc

```diff
@@ -140,6 +140,154 @@ SyncRes::SyncRes(const struct timeval& now) :  d_authzonequeries(0), d_outquerie
 {
 }
 
+static void allowAdditionalEntry(std::unordered_set<DNSName>& allowedAdditionals, const DNSRecord& rec);
+
+void SyncRes::resolveAdditionals(const DNSName& qname, QType qtype, AdditionalMode mode, std::vector<DNSRecord>& additionals, unsigned int depth)
+{
+  vector<DNSRecord> addRecords;
+
+  vState state = vState::Indeterminate;
+  switch (mode) {
+  case AdditionalMode::ResolveImmediately: {
+    set<GetBestNSAnswer> beenthere;
+    int res = doResolve(qname, qtype, addRecords, depth, beenthere, state);
+    if (res != 0) {
+      return;
+    }
+    // We're conservative here. We do not add Bogus records in any circumstance, we add Indeterminates only if no
+    // validation is required.
+    if (vStateIsBogus(state)) {
+      return;
+    }
+    if (shouldValidate() && state != vState::Secure && state != vState::Insecure) {
+      return;
+    }
+    for (const auto& rec : addRecords) {
+      if (rec.d_place == DNSResourceRecord::ANSWER) {
+        additionals.push_back(rec);
+      }
+    }
+    break;
+  }
+  case AdditionalMode::CacheOnly:
+  case AdditionalMode::CacheOnlyRequireAuth: {
+    // Peek into cache
+    if (g_recCache->get(d_now.tv_sec, qname, qtype, mode == AdditionalMode::CacheOnlyRequireAuth, &addRecords, d_cacheRemote, false, d_routingTag, nullptr, nullptr, nullptr, &state) <= 0) {
+      return;
+    }
+    // See the comment for the ResolveImmediately case
+    if (vStateIsBogus(state)) {
+      return;
+    }
+    if (shouldValidate() && state != vState::Secure && state != vState::Insecure) {
+      return;
+    }
+    for (auto& rec : addRecords) {
+      if (rec.d_place == DNSResourceRecord::ANSWER) {
```

Comment:
I believe this is always true?

Reply:
Indeed, but I like to leave it here though for symmetry with the `ResolveImmediate` case where it is needed.